ext/opcache/jit/zend_jit_ir.c

```diff
@@ -14024,6 +14024,52 @@ static int zend_jit_class_guard(zend_jit_ctx *jit, const zend_op *opline, ir_ref
 	return 1;
 }
 
+/* Child classes that add new hooks to existing properties will violate the
+ * assumption that properties may be accessed directly. This guard makes sure no
+ * new hooked properties have been created in the object, compared to the trace
+ * class. For function JIT, we just pick the slow path. */
+static bool zend_jit_new_hooks_guard(
+	zend_jit_ctx *jit,
+	const zend_op *opline,
+	zend_ssa *ssa,
+	const zend_ssa_op *ssa_op,
+	ir_ref obj_ref,
+	uint32_t op1_info,
+	zend_class_entry *ce,
+	bool ce_is_instanceof,
+	zend_property_info *prop_info,
+	ir_ref *slow_inputs
+) {
+	if ((JIT_G(trigger) != ZEND_JIT_ON_HOT_TRACE || !(op1_info & MAY_BE_NEW_HOOKS_GUARD))
+	 && ce_is_instanceof
+	 && !(prop_info->flags & ZEND_ACC_FINAL)) {
+		ir_ref num_hooked_props = ir_LOAD_A(ir_ADD_OFFSET(obj_ref, offsetof(zend_object, ce)));
+		num_hooked_props = ir_LOAD_U32(ir_ADD_OFFSET(num_hooked_props, offsetof(zend_class_entry, num_hooked_props)));
+		ir_ref has_same_num_hooked_props = ir_EQ(num_hooked_props, ir_CONST_U32(ce->num_hooked_props));
```

Comment:
If I'm not mistaken, you can convert hooks back to normal properties. So is it not possible to trick this check in that way?

Reply:
Hooks cannot be discarded, they are always inherited. Even if, the hooks trigger the slow path, which would still work without hooks. At least if I understood your comment.

Reply:
Then why does this work? https://3v4l.org/tSDaN

Reply:
I feel like I need to re-read the RFC again... I didn't get that from intuition.
So it appears indeed that the parent hook is called, but then this is allowed which looks like a bug? https://3v4l.org/2sFXl

Reply:
A backed property can't go back to virtual, the slot is there and might be accessed from a parent hook. Think of it like a getProp method, it's always inherited, even if you were to redeclare the property itself.

Reply:
I see, thanks. I'll re-read it again as it's been a while; and experiment a bit with it.

Reply:
Thanks! The inheritance semantics are indeed not obvious,  and it took us a while to find some that we considered to have the best tradeoffs.